keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,1069 @@
+    ... # omitted to save space
+```
+
+Similarly `kube-controller-manager`'s manifest would like:-
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  creationTimestamp: null
+  labels:
+    component: kube-controller-manager
+    tier: control-plane
+  name: kube-controller-manager
+  namespace: kube-system
+spec:
+  securityContext:
+    supplementalGroups:
+    - 2100
+  containers:
+  - name: kube-controller-manager
+    command:
+    - kube-controller-manager
+    - --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt
+    ...
+    # omitted to save space.
+    securityContext:
+      runAsUser: 2001  # this value is only an example and is not the id we plan to use.
+      runAsGroup: 2001 # this value is only an example and is not the id we plan to use.
+      allowPrivilegeEscalation: false
+      seccompProfile:
+        type: runtime/default
+      capabilities:
+        drop:
+        - ALL
+    image: k8s.gcr.io/kube-controller-manager:v1.21.0-beta.0.368_9850bf06b571d5-dirty
+    ... # omitted to save space.
+    volumeMounts:
+    - mountPath: /etc/kubernetes/pki
+      name: k8s-certs
+      readOnly: true
+    ... # omitted to save space. 
+```
+
+Here is a table of all the things that `kube-apiserver`, `kube-controller-manager`, `kube-scheduler` and `etcd` mount. 
+
+World readable stuff:
+- /usr/local/share/ca-certificates
+- /usr/share/ca-certificates
+- /etc/ssl/certs
+- /etc/ca-certificates
+
+<br />
+
+**Table key:-**
+
+| Abbreviation | Explanation |
+|--------------|-------------|
+| etcd | The UID/GID that we will assign to `etcd` |
+| kas  | The UID/GID that we will assign to `kube-apiserver` |
+| kcm | The UID/GID that we will assign to `kube-controller-manager` |
+| ks | The UID/GID that we will assign to `kube-scheduler` |
+| etcd_ca_crt_readers | The GID we will assign to a group that allows you to read /etc/kubernetes/pki/etcd/ca.crt |
+| ca_crt_readers | The GID we will assign to a group that allows you to read /etc/kubernetes/pki/etcd/ca.crt |
+| sa_key_readers | The GID we will assign to a group that allows you to read /etc/kubernetes/pki/sa.key |
+| front_proxy_ca_crt_readers | The GID we will assign to a group that allows you to read /etc/kubernetes/pki/front-proxy-ca.crt |
+
+<br/>
+
+**Stuff that we care about for this kep:-**
+| file/directory                                   | Components | File permission |
+| -------------------------------------------------|------------|-----------------|
+| /etc/kubernetes/pki/etcd/server.crt              | etcd       | 640 etcd etcd   |
+| /etc/kubernetes/pki/etcd/server.key              | etcd       | 640 etcd etcd   |
+| /etc/kubernetes/pki/etcd/peer.crt                | etcd       | 640 etcd etcd   |
+| /etc/kubernetes/pki/etcd/peer.key                | etcd       | 640 etcd etcd   |
+| /etc/kubernetes/pki/etcd/ca.crt                  | etcd, kas  | 640 root etcd_ca_crt_readers |
+| /var/lib/etcd/                                   | etcd       | 640 etcd etcd   |
```

Comment:
note: this is an interesting one for me and just wanted to highlight it for discussion.

currently both /var/lib/kubelet and /var/lib/etcd are root owned and hopefully this change would be OK.
currently kubeadm creates the /var/lib/etcd folder before it writes the static pod manifests.

Reply:
Yeah I think that totally fine. We would need to ensure that we create the uid/gids for all components before we create any files or directories and when we create the files/directories the right permissions are set.

In the case of upgrade this would simply be a chown -R /var/lib/etcd/ that way when we update the etcd pod, it will be able to read/write both old and new data.